src/Chapter4.hs

```diff
@@ -488,7 +503,17 @@ Implement the 'Applicative' instance for our 'List' type.
   may also need to implement a few useful helper functions for our List
   type.
 -}
+instance Applicative List where
+  pure :: a -> List a
+  pure x = Cons x Empty
+
+  (<*>) :: List (a -> b) -> List a -> List b
+  (<*>) Empty _ = Empty
+  (<*>) (Cons f fs) xs = concatList (fmap f xs) ((<*>) fs xs)
 
+concatList :: List a -> List a -> List a
+concatList Empty ys = ys
```

Comment:
Perfect!

You can add one more case to make your solution even more efficient:

```haskell
append xs Empty = xs
```

In that case, you won't need to walk through each element of the first list when the second one is empty and return the list instantly 🚤

Reply:
Oh, I see. I'll do it then.
```suggestion
concatList Empty ys = ys
concatList ys Empty = ys
```